src/WebJobs.Script/Host/FunctionMetadataManager.cs

```diff
@@ -18,36 +22,91 @@ namespace Microsoft.Azure.WebJobs.Script
-        private readonly ILogger _logger;
+        private bool _isHttpWorker;
+        private bool _servicesReset = false;
+        private ILogger _logger;
+        private IOptions<ScriptJobHostOptions> _scriptOptions;
+        private ImmutableArray<FunctionMetadata> _functionMetadataFilteredArray;
+        private ImmutableArray<FunctionMetadata> _functionMetadataAllArray;
+        private IEnumerable<IFunctionProvider> _functionProviders;
         private Dictionary<string, ICollection<string>> _functionErrors = new Dictionary<string, ICollection<string>>();
 
-        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IOptions<HttpWorkerOptions> httpWorkerOptions, ILoggerFactory loggerFactory)
+        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IEnumerable<IFunctionProvider> functionProviders, IOptions<HttpWorkerOptions> httpWorkerOptions, IScriptHostManager scriptHostManager, ILoggerFactory loggerFactory)
         {
             _scriptOptions = scriptOptions;
+            _serviceProvider = scriptHostManager as IServiceProvider;
             _functionMetadataProvider = functionMetadataProvider;
             _logger = loggerFactory.CreateLogger(LogCategories.Startup);
-            _functionMetadataArray = new Lazy<ImmutableArray<FunctionMetadata>>(LoadFunctionMetadata);
-            _isHttpWorker = httpWorkerOptions.Value.Description != null;
-        }
+            _isHttpWorker = httpWorkerOptions?.Value?.Description != null;
+            _functionProviders = functionProviders;
 
-        public ImmutableArray<FunctionMetadata> Functions => _functionMetadataArray.Value;
+            scriptHostManager.HostInitializing += ResetScriptHostServices;
+        }
 
         public ImmutableDictionary<string, ImmutableArray<string>> Errors { get; private set; }
 
+        /// <summary>
+        /// Gets the function metadata array from all providers.
+        /// </summary>
+        /// <param name="forceRefresh">Forces reload from all providers.</param>
+        /// <param name="includeBlocked">Includes all functions, including blocked ones.</param>
+        /// <returns>An Immmutable array of FunctionMetadata.</returns>
+        public ImmutableArray<FunctionMetadata> GetFunctionMetadata(bool forceRefresh, bool includeBlocked = false)
+        {
+            var currentlyLoaded = GetLoadedMetadata(includeBlocked);
+            if (!forceRefresh && !_servicesReset && !currentlyLoaded.IsDefaultOrEmpty)
+            {
+                return currentlyLoaded;
+            }
+
+            LoadFunctionMetadata(forceRefresh);
+            _servicesReset = false;
+
+            return GetLoadedMetadata(includeBlocked);
+        }
+
+        internal ImmutableArray<FunctionMetadata> GetLoadedMetadata(bool includeBlocked)
```

Comment:
Let's remove this GetLoadedMetadata function as well as the two _functionMetadataAllArray/_functionMetadataFilteredArray members, and just do the filtering herebefore returning. It's just a simple filters over a small collection - not a perf issue, given this API isn't called on hot paths. To handle your includeBlocked param, just do the whitelist filter here. Rename the param to applyWhitelist = true - I'd rather not have a new "blocked" alias with the existing "whitelist" concept.